### ISSUE TITLE
AC-10 mitigation (concurrent sessions) (V16-AC-10)

### ISSUE THREAD

**Issue by mogul** [18F/cg-product#559]
In order to mitigate risk from our lack of implementation for NIST 800-53 control AC-10, we want to ensure unusually high concurrent session numbers result in alerts to operators.

**Acceptance criteria**

[ ] We complete and upload a [deviation request form](https://drive.google.com/file/d/0B1cewEqKcWCbU1hUSHNreng2SW8/view) as part of the the February ConMon report indicating our intention to wait for the upstream implementation, and mail the FedRAMP PMO with a pointer to it.

**OLD Acceptance Criteria**

- ~~If an account has more than 3 active sessions via opslogin UAA, an alert is sent~~
- ~~If an admin account has more than 3 active sessions via customer UAA, an alert is sent~~
- ~~If a non-admin account has more than 2 active sessions via customer UAA, an alert is sent~~

**Implementation notes**

- ~~Change logs.cloud.gov to refer to the admin scope, rather than a space role, to determine who's an administrator~~ (Done)
- Write a script that counts up the number of tokens for each user in [opslogin|customer] UAA with admin scope and emits that number to Riemann where we alert if > 3
- Write a script that counts up the number of tokens for each user in customer UAA without admin scope and emits that number to Riemann where we alert if > 2

**Notes**

Reference [comments from the AO on our justifications](https://docs.google.com/document/d/1sMVT-lcgHzzNvsvxqpEJ3OY04WXwAQhY5ObA0gIPrzg/edit#).

**Comment by cnelson** [comment]
We should also figure out what our requirements would be if UAA had session management, so we can communicate them back to the UAA devs.  They seemed receptive to adding concurrent session control to their backlog last time I spoke with them about it.

**Comment by mogul** [comment]
The [NIST guidance](https://web.nvd.nist.gov/view/800-53/Rev4/control?controlName=AC-10) doesn't say how many concurrent sessions should/shouldn't be allowed, only suggests that the number of sessions should be tracked and limited. It doesn't even suggest the specific goal or risk being mitigated by having this control!

I think to be useful for us, we'd want to specify how many concurrent sessions are allowable. For example, we might determine that operators typically have one web-based session and one CLI session active at a time when working on an average day, and rarely if ever need a third session (eg from another machine). So I think we'd want to the ability to set the limit to two, and have the oldest session ended when that number is exceeded. What do you think?

(Alternatively, I've seen in some SaaS that you get a banner at the top of the screen when you have multiple sessions active, saying "You have X other sessions active. Click _here_ for details or _here_ to log them out." I don't think that's all that useful here since we have CLI sessions as well, so the banner won't typically be seen.)

**Comment by mogul** [comment]
(Plus it'd be difficult to implement across all of the relying-parties for UAA.)

**Comment by cnelson** [comment]
> we might determine that operators typically have one web-based session and one CLI session active at a time when working on an average day, and rarely if ever need a third session

I think it's important to note that a session != a token in UAA.  A session just means that you aren't prompted to authenticate again when authorizing applications, getting one-time passcodes etc.

Once you've `cf login --sso` for example to establish a CLI session, UAA isn't keeping a "session" alive for you, it's just issued a token with a defined lifetime, and the cf client uses it to perform actions on your behalf.

If the goal is to make visible which applications have *active tokens* we could do that today by  updating [uaa-extras](https://github.com/18F/cg-uaa-extras) to use the already available APIs (https://docs.cloudfoundry.org/api/uaa/#list-tokens , https://docs.cloudfoundry.org/api/uaa/#revoke-tokens) to allow users to inspect and revoke any active tokens.

We could still start a discussion about adding concurrent session control to UAA with Pivotal but I think it'd be a lot less important (maybe not needed?) if we gave users the ability to manage their tokens.

Thoughts?

cc: @NoahKunin

**Comment by mogul** [comment]
I think tokens ~= sessions for our purposes here. Providing explicit user-management of those tokens is a nice-to-have thing, but it's really not needed for compliance AFIACS. 

I think we're good if we just have something that revokes the oldest tokens to reach the "no more than N" threshold. What do you think?

**Comment by mogul** [comment]
I said above:

> The [NIST guidance](https://web.nvd.nist.gov/view/800-53/Rev4/control?controlName=AC-10) doesn't say how many concurrent sessions should/shouldn't be allowed, only suggests that the number of sessions should be tracked and limited. It doesn't even suggest the specific goal or risk being mitigated by having this control!

But as we found in our [further discussion in Slack](https://gsa-tts.slack.com/archives/cloud-gov-highbar/p1482345453000830) FedRAMP is more explicit in their assignment:

> The information system limits the number of concurrent sessions for each [Assignment: organization-defined account and/or account type] to [FedRAMP Assignment: three (3) sessions for privileged access and two (2) sessions for non-privileged access].

As discussed with our AO for the POAM, we're just going to do alerting for now and see what data we can gather from that about how disruptive a hard restriction would be. I've edited the initial post to reflect this story as groomed in that direction.

(Further work to expose token listing and revocation to users is a nice-to-have for a future time,  but it's not in scope here until we see a demonstrated user need for it.)

**Comment by jmcarp** [comment]
I moved @LinuxBozo 's token audit script to cg-deploy-monitoring as a cron job to periodically send counts to riemann. We also updated the uaa credentials broker to set token expiration. We're going to check out the token counts in influxdb next week to understand how to set alerting thresholds--although it's already clear that many users have more than two or three tokens at a time. We also have a wippr at https://github.com/18F/cg-riemann-boshrelease/pull/64 for alerting that we're holding off on for now.

@mogul: should we move this back to ready or grooming until we have enough data in influx for alerting, or close and make a second story, or do something else?

**Comment by cnelson** [comment]
Can we document the influx query / command to run to get these counts by day?

**Comment by jmcarp** [comment]
Not sure of the most useful query / queries to look at, but here's a start:

```
influxdb/492997b6-1444-4bc4-8a87-c60274d7792b:~# curl -G 'http://localhost:8086/query?pretty=true' --data-urlencode "db=production" --data-urlencode 'q=select perentile("value", 10) as p10, percentile("value", 50) as p50, percentile("value", 90) as p90 from "
uaa-token-audit" where time > (now() - 5d) group by time(1d)'
{
    "results": [
        {
            "series": [
                {
                    "name": "uaa-token-audit",
                    "columns": [
                        "time",
                        "p10",
                        "p50",
                        "p90"
                    ],
                    "values": [
                        [
                            "2017-01-27T00:00:00Z",
                            1,
                            1,
                            11
                        ],
                        [
                            "2017-01-28T00:00:00Z",
                            1,
                            2,
                            12
                        ],
                        [
                            "2017-01-29T00:00:00Z",
                            1,
                            1,
                            9
                        ],
                        [
                            "2017-01-30T00:00:00Z",
                            1,
                            1,
                            9
                        ],
                        [
                            "2017-01-31T00:00:00Z",
                            1,
                            1,
                            8
                        ],
                        [
                            "2017-02-01T00:00:00Z",
                            1,
                            1,
                            9
                        ]
                    ]
                }
            ]
        }
    ]
}
```

**Comment by mogul** [comment]
We need to look at our data once our current secret rotation effort is complete to determine what else to do here.

**Comment by rememberlenny** [comment]
Updates

Reviewed logged params and added user's email.
Discovered the current expire time is 30 days, so many users have persisting tokens.
Identifying the average range or token expiration method we should implement.

**Comment by rememberlenny** [comment]
[@jmcarp - Josh Carp] Fixed the expire tokens yesterday to be 3 days.

**Comment by rememberlenny** [comment]
https://cloudfoundry.slack.com/archives/uaa/p1487280330001967

Requesting if UAA will integrate the token limit.

Backup requests:

- Control the number of active tokens
- Request the current active tokens, sorted by time
- Delete multiple tokens in a single request
- Check the status of multiple tokens in a single request
- Only return tokens that are active (not-revoked) from a user-id

**Comment by rememberlenny** [comment]
Current status:

We need clarity from FedRamp requirements around number of tokens.

Question for [@brittag - Britta Gustafson] [@mogul - Bret Mogilefsky] : Are the tokens per user OR tokens per user on a client. For example, Person gets two tokens across all the applications they are authenticated on? Or do they get two authenticated tokens, per application that they need to be authenticated into.

Please note, option 1 is not possible.

--

Ideal outcomes based on result:

1. UAA handles everything.
2. UAA is okay with feature. We write the PR (java). It gets accepted.
3. We make a database trigger that checks for the number of sessions on a user-id per client-id. Based on exceeding certain threshold, trigger will remove oldest session accordingly.

**Comment by brittag** [comment]
I believe tokens per user on a client should be ok. References, for context:

* FedRAMP rqeuirement: _The information system limits the number of concurrent sessions for each [Assignment: organization-defined account and/or account type] to [FedRAMP Assignment: three (3) sessions for privileged access and two (2) sessions for non-privileged access]._
* "Supplemental Guidance" at https://web.nvd.nist.gov/view/800-53/Rev4/control?controlName=AC-10

**Comment by rememberlenny** [comment]
Still waiting on Cloud Foundry on UAA feature request.

**Comment by cnelson** [comment]
Is waiting for upstream our answer for this?

If we don't intend to solve this temporarily with our own job to prune tokens while we wait on upstream can we close this issue?

**Comment by jmcarp** [comment]
If waiting for upstream is acceptable, then yes, I think we should close this. I just pinged upstream asking for details at https://github.com/cloudfoundry/uaa/issues/540#issuecomment-281788536.

**Comment by mogul** [comment]
We're going to document our plan to wait for the upstream fix, ACs updated accordingly.

**Comment by brittag** [comment]
We started the Deviation Request Form here: https://docs.google.com/document/d/1EbE0jze4pPYczuGOYD6CNWngXu3e9diISOnIOwawfxE/edit#heading=h.gjdgxs

**Comment by brittag** [comment]
The next step is for Lenny to put in a few more technical details and then Britta will polish the writing.

**Comment by brittag** [comment]
This Deviation Request Form at https://docs.google.com/document/d/1EbE0jze4pPYczuGOYD6CNWngXu3e9diISOnIOwawfxE/edit# still needs: request number (Britta will figure that out upon submitting it), and a bit more writing to explain our planned remediation and justify why we're doing it.

**Comment by rememberlenny** [comment]
Added updates to the Deviation Request Form

**Comment by brittag** [comment]
Followup work is tracked at https://favro.com/card/1e11108a2da81e3bd7153a7a/18F-3556
